File: src/scpn_phase_orchestrator/supervisor/events.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Phase Orchestrator — Supervisor event types

"""Validated supervisor event records and an in-process bounded event bus.

``RegimeEvent`` restricts event kinds and step/detail fields before publication,
and ``EventBus`` records a bounded chronological history while notifying
callable subscribers synchronously. The bus is process-local and passive: it
does not spawn threads, persist logs, retry subscriber failures, or emit network
traffic.
"""

from __future__ import annotations

from collections import deque
from collections.abc import Callable
from dataclasses import dataclass
from numbers import Integral

__all__ = ["RegimeEvent", "EventBus"]

VALID_EVENT_KINDS = frozenset(
    {
        "boundary_breach",
        "r_threshold",
        "regime_transition",
        "manual",
        "petri_transition",
    }
)


def _validate_nonnegative_int(value: object, *, name: str) -> int:
    if isinstance(value, bool) or not isinstance(value, Integral) or value < 0:
        raise ValueError(f"{name} must be a non-negative integer, got {value!r}")
    return int(value)


def _validate_positive_int(value: object, *, name: str) -> int:
    if isinstance(value, bool) or not isinstance(value, Integral) or value < 1:
        raise ValueError(f"{name} must be >= 1, got {value!r}")
    return int(value)


@dataclass(frozen=True)
class RegimeEvent:
    """Immutable event emitted on regime transitions or boundary breaches."""

    kind: str
    step: int
    detail: str = ""

    def __post_init__(self) -> None:
        if self.kind not in VALID_EVENT_KINDS:
            raise ValueError(
                f"invalid event kind {self.kind!r}, "
                f"expected one of {sorted(VALID_EVENT_KINDS)}"
            )
        _validate_nonnegative_int(self.step, name="step")
        if not isinstance(self.detail, str):
            raise ValueError(f"detail must be a string, got {self.detail!r}")


class EventBus:
    """Pub/sub bus for regime events with bounded history."""

    def __init__(self, maxlen: int = 200) -> None:
        maxlen = _validate_positive_int(maxlen, name="maxlen")
        self._subscribers: list[Callable[[RegimeEvent], None]] = []
        self._history: deque[RegimeEvent] = deque(maxlen=maxlen)

    def subscribe(self, callback: object) -> None:
        """Register a callback to receive future events."""
        if not callable(callback):
            raise ValueError(f"callback must be callable, got {callback!r}")
        self._subscribers.append(callback)

    def unsubscribe(self, callback: object) -> None:
        """Remove a previously registered callback."""
        self._subscribers = [s for s in self._subscribers if s != callback]

    def post(self, event: RegimeEvent) -> None:
        """Record *event* in history and notify all subscribers."""
        if not isinstance(event, RegimeEvent):
            raise ValueError(f"event must be a RegimeEvent, got {event!r}")
        self._history.append(event)
        for cb in tuple(self._subscribers):
            cb(event)

    @property
    def history(self) -> list[RegimeEvent]:
        """Chronological list of all posted events."""
        return list(self._history)

    @property
    def count(self) -> int:
        """Number of events in history."""
        return len(self._history)

    def clear(self) -> None:
        """Discard all recorded events."""
        self._history.clear()

File: spo-kernel/crates/spo-supervisor/src/events.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Phase Orchestrator — Supervisor event types

use std::collections::VecDeque;

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum EventKind {
    BoundaryBreach,
    RThreshold,
    RegimeTransition,
    Manual,
    PetriTransition,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RegimeEvent {
    pub kind: EventKind,
    pub step: u64,
    pub detail: String,
}

impl RegimeEvent {
    #[must_use]
    pub fn new(kind: EventKind, step: u64, detail: String) -> Self {
        Self { kind, step, detail }
    }
}

/// Bounded event history. Callbacks are handled on the Python side;
/// the Rust EventBus is a pure append-only ring buffer for audit replay.
#[derive(Debug)]
pub struct EventBus {
    history: VecDeque<RegimeEvent>,
    max_history: usize,
}

impl EventBus {
    #[must_use]
    pub fn new(max_history: usize) -> Self {
        Self {
            history: VecDeque::with_capacity(max_history),
            max_history,
        }
    }

    pub fn post(&mut self, event: RegimeEvent) {
        if self.history.len() == self.max_history {
            self.history.pop_front();
        }
        self.history.push_back(event);
    }

    #[must_use]
    pub fn history(&self) -> &VecDeque<RegimeEvent> {
        &self.history
    }

    #[must_use]
    pub fn count(&self) -> usize {
        self.history.len()
    }

    pub fn clear(&mut self) {
        self.history.clear();
    }
}

impl Default for EventBus {
    fn default() -> Self {
        Self::new(200)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn event_creation() {
        let e = RegimeEvent::new(EventKind::RegimeTransition, 5, "a->b".into());
        assert_eq!(e.kind, EventKind::RegimeTransition);
        assert_eq!(e.step, 5);
        assert_eq!(e.detail, "a->b");
    }

    #[test]
    fn bus_post_and_history() {
        let mut bus = EventBus::new(10);
        bus.post(RegimeEvent::new(EventKind::Manual, 1, String::new()));
        bus.post(RegimeEvent::new(EventKind::Manual, 2, String::new()));
        assert_eq!(bus.count(), 2);
    }

    #[test]
    fn bus_bounded_history() {
        let mut bus = EventBus::new(3);
        for i in 0..5 {
            bus.post(RegimeEvent::new(EventKind::Manual, i, String::new()));
        }
        assert_eq!(bus.count(), 3);
        assert_eq!(bus.history().front().expect("not empty").step, 2);
    }

    #[test]
    fn bus_clear() {
        let mut bus = EventBus::default();
        bus.post(RegimeEvent::new(EventKind::Manual, 1, String::new()));
        bus.clear();
        assert_eq!(bus.count(), 0);
    }

    #[test]
    fn event_serde_roundtrip() {
        let e = RegimeEvent::new(EventKind::BoundaryBreach, 42, "test".into());
        let json = serde_json::to_string(&e).expect("serialize");
        let e2: RegimeEvent = serde_json::from_str(&json).expect("deserialize");
        assert_eq!(e, e2);
    }
}
